src/client/session/action.rs

```diff
@@ -0,0 +1,400 @@
+    ///     }.boxed()
+    /// ).await?;
+    /// /* This will not compile with a "variable moved due to use in generator" error:
+    /// session.start_transaction().and_run(
+    ///     (),
+    ///     |session, _| async move {
+    ///         coll.insert_one(doc! { "data": my_data }).session(session).await
+    ///     }.boxed()
+    /// ).await?;
+    /// */
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn and_run<R, C, F>(self, mut context: C, mut callback: F) -> Result<R>
+    where
+        F: for<'b> FnMut(&'b mut ClientSession, &'b mut C) -> BoxFuture<'b, Result<R>>,
+    {
+        let timeout = Duration::from_secs(120);
+        #[cfg(test)]
+        let timeout = self
+            .session
+            .0
+            .convenient_transaction_timeout
+            .unwrap_or(timeout);
+        let start = Instant::now();
+
+        use crate::error::{TRANSIENT_TRANSACTION_ERROR, UNKNOWN_TRANSACTION_COMMIT_RESULT};
+
+        'transaction: loop {
+            self.session
+                .0
+                .start_transaction()
+                .with_options(self.options.clone())
+                .await?;
+            let ret = match callback(self.session.0, &mut context).await {
+                Ok(v) => v,
+                Err(e) => {
+                    if matches!(
+                        self.session.0.transaction.state,
+                        TransactionState::Starting | TransactionState::InProgress
+                    ) {
+                        self.session.0.abort_transaction().await?;
+                    }
+                    if e.contains_label(TRANSIENT_TRANSACTION_ERROR) && start.elapsed() < timeout {
+                        continue 'transaction;
+                    }
+                    return Err(e);
+                }
+            };
+            if matches!(
+                self.session.0.transaction.state,
+                TransactionState::None
+                    | TransactionState::Aborted
+                    | TransactionState::Committed { .. }
+            ) {
+                return Ok(ret);
+            }
+            'commit: loop {
+                match self.session.0.commit_transaction().await {
+                    Ok(()) => return Ok(ret),
+                    Err(e) => {
+                        if e.is_max_time_ms_expired_error() || start.elapsed() >= timeout {
+                            return Err(e);
+                        }
+                        if e.contains_label(UNKNOWN_TRANSACTION_COMMIT_RESULT) {
+                            continue 'commit;
+                        }
+                        if e.contains_label(TRANSIENT_TRANSACTION_ERROR) {
+                            continue 'transaction;
+                        }
+                        return Err(e);
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[cfg(feature = "sync")]
+impl<'a> StartTransaction<SyncSession<'a>> {
+    /// Synchronously execute this action.
```

Comment:
I implemented this manually rather than having an `impl<'a> Action for StartTransaction<SyncSession<'a>>` because it doesn't make much sense to have a distinct future type or `IntoFuture` impl for the sync version.